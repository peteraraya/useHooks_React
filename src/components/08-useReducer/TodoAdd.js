import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
// uso de custom hook , este deberia encargarse de limpiar los campos 
const [{ description }, handleInputChange, reset] = useForm({
    // tendremos un objeto que tendra la llave de la description , lo vamos asociar con el description del inpit (deben tener el mismo nombre )
    description: ''
});

const handleSubmit = (e) =>{
    e.preventDefault();

    if (description.trim().length <= 1) {
        return;
    }

    const newTodo = {
        id: new Date().getTime(),
        desc: description,
        done: false
    };


    handleAddTodo( newTodo );
    reset();

}


    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    className="form-control"
                    name="description"
                    placeholder="Aprender..."
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />
                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                        </button>
            </form>
        </>
    )
}

