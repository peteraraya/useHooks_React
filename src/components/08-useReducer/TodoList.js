import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul className="list-group listo-group-flush">
            {
                todos.map((todo, i) => (
                 <TodoListItem 
                        key = { todo.id }
                        todo = { todo } 
                        index =  { i } 
                        handleDelete = { handleDelete }
                        handleToggle={handleToggle }
                 />
                ))
            }
        </ul>
    )
}


/**
 * el indice nunca debemos ocuparlo de index
 *   */ 
