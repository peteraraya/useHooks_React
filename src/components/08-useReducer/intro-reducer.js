
const initialState = [{
    id: 1,
    todo: 'Comprar Leche',
    done: false
}];


// Puede ser considerado un reducer

const todoReducer = (state = initialState, action) => {

    // ya que solo me esta devolviendo el estado anterior agrego un if, y agregamos ? en el action en caso de error por undefined
    if (action?.type === 'agregar') {

        return [...state, action.payload]
    }

    return state
}
let todos = todoReducer();



// a침adir un nuevo elemento a un todo
// no utilizar push

const newTodo = {
    id: 2,
    todo: 'Comprar Leche',
    done: false
}

// A traves de la acci칩n enviamos m치s informaci칩n

const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo
}

todos = todoReducer(todos, agregarTodoAction);

console.log(todos)

