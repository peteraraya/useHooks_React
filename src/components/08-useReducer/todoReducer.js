
export const todoReducer = (state = [] , action) => {

    switch ( action.type ) {
        case 'add':
            
            return [ ...state, action.payload];

        case 'delete':

            return state.filter( todo => todo.id !==  action.payload ); 
    

        case 'toogle':

            return state.map( todo =>
                ( todo.id === action.payload )
                    ? { ...todo, done: !todo.done } // retorna modificación
                    : todo // en caso de no hacer nada
            );
        

        case 'toogle-old':
            return state.map( todo => {
                // evaluo 
                if ( todo.id  === action.payload ) {
                    // este elemento es el que hay que cambiar
                    return {

                        ...todo,
                        done: !todo.done

                    }
                }else{
                    return todo;
                }
            })

        default:
            return state;
    }


}

// filter regresa un nuevo arreglo que cumplan una condición
// map : barre cada uno de los elementos del arreglo y debe de retornar un valor
