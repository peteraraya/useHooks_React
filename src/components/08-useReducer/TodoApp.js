import React, { useReducer, useEffect } from 'react';
import { todoReducer } from './todoReducer';
import './style.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

// const initialState = [{
//     id: new Date().getTime(),
//     desc: 'Aprende React',
//     donae: false
// }];

// Función init
const init = () => {
    // return[{
    //     id: new Date().getTime(),
    //     desc: 'Aprende React',
    //     donae: false
    // }];

    return JSON.parse(localStorage.getItem('todos')) || [];
    // Si existen todos: regresos esos todos como arreglos
    //  si no existen debo regresar un arreglo vació que va ser mi estado inicial

}


export const TodoApp = () => {

    // reducer      : tiene que ser la función reducer
    // initialState : es un estado incial que podemos mandar
    // init         : es usado como una función para inicializar el state em caso que el state sea como procesado o haga varias acciones, puede memorizar el estado

    // dispatch : ayudar a disparar las acciones y va a saber a quereducer ahay que mandar la información y cuando cambie el state va a redibujar lo que cambie
    // a este se le envia la aaccion

    // función init nos va ayudar  a computar todo el estado inicial para que funcione más rapido el componente y esa función no se esté ejecutando a cada vez que hay cambios


    const [todos, dispatch] = useReducer(todoReducer, [], init);


    // conectamos nuestro customhook con un onChange y debo asignarle el value
    // Grabar en localstorage que solo guarda strings

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos]);


    const handleDelete = ( todoId ) => {

        // console.log( todoId );

        // creamos accion
        const action = {
            type: 'delete',
            payload: todoId
        }

        // hacer el dispath
        dispatch( action );

    }

    const handleToggle = ( todoId ) =>{
        dispatch({
            type:'toggle',
            payload: todoId
        });
    }

    const handleAddTodo = ( newTodo ) =>{

        dispatch({
            type: 'add',
            payload: newTodo
        });
    }


    return (
        <div>
            <h1>TodoApp ({todos.length}) </h1>
            <hr />
            <div className="row">

                <div className="col-7">
                   
                   <TodoList
                        // proptypes
                        todos={ todos }
                        handleDelete = { handleDelete }
                        handleToggle = { handleToggle }
                   />
                </div>
                <div className="col-5">
                    
                   <TodoAdd
                        handleAddTodo={ handleAddTodo }
                   />
                </div>

            </div>
        </div>
    )
}



// Grabar información en localstorage - cuando los todos cambien
// usaremos useEffect


// el efecto se va encargar de guardar cualquier cambio que exista en esos todos (agrege, modifique lo va grabar en locastorage)
