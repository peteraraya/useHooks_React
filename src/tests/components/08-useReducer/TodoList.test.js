import React from 'react';

import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../../fixtures/demoTodos';

describe('Pruebas en <TodoLis />', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    
    const wrapper = shallow(
        <TodoList 
            todos = { demoTodos }
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
        />)
    
    test('debe de mostrase correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('debe de tener dos o cantidad de elementos que tenga <TodoListItem />', () => {
       console.log( wrapper.find('TodoListItem').length );
        expect( wrapper.find('TodoListItem').length ).toBe( demoTodos.length );

        console.log(wrapper.find('TodoListItem').at(0).prop('handleDelete'));

        expect( wrapper.find('TodoListItem').at(0).prop('handleDelete') ).toEqual( expect.any(Function) );
    });

    test('debe de mostrase correctamente', () => {
       
    });


})
