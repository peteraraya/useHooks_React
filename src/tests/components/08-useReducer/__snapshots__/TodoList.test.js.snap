// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoLis /> debe de mostrase correctamente 1`] = `
<ul
  className="list-group listo-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React 2",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Angular",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
